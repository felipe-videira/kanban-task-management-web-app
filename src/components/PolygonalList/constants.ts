export const SCALE_ROTATION_45DEG = 1.4142;
export const POINTING_UP_ROTATION = 45;
export const POINTING_DOWN_ROTATION = 225;
