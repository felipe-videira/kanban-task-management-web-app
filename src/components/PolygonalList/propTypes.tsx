import { elementType, number, object, arrayOf, bool } from "prop-types";

export default {
  data: arrayOf(object).isRequired,
  ItemComponent: elementType.isRequired,
  itemProps: object,
  itemSize: number.isRequired,
  pointingUp: bool,
};
