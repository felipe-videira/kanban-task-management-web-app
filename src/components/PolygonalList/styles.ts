import styled from "styled-components/macro";

const SCALE_ROTATION_45DEG = 1.4142;

interface ContainerProps {
  readonly size: number;
  readonly pointingUp?: boolean | null;
}

export const PolygonalListContainer = styled.div.attrs<ContainerProps>(
  (props) => ({
    size: `${props.size / SCALE_ROTATION_45DEG}px`,
  })
)<ContainerProps>`
  ${(props) => `
    transform: rotate(${props.pointingUp ? "45" : "225"}deg);
    width: ${props.size};
    height: ${props.size};
`}
`;

interface ItemContainerProps {
  readonly length: number;
  readonly itemSize: number;
  readonly zIndex: number;
  readonly origin?: string;
}

export const PolygonalListItemContainer = styled.div.attrs<ItemContainerProps>(
  (props) => ({
    style: {
      zIndex: props.zIndex,
      transform: `rotate(${360 / props.length}deg)`,
    },
    origin: `-${props.itemSize / (props.itemSize * 0.1)}px`,
  })
)<ItemContainerProps>`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  ${(props) =>
    props.length >= 3 &&
    `
      &::after {
          content: "";
          position: absolute;
          transform-origin: ${props.origin} ${props.origin};
          height: calc(100% / (${props.length}/ 4));
          width: ${props.itemSize * 0.1}px;
          background: rgba(0,0,0,0.2);
          top: 0;
          left: 0;
          transform: rotate(${(90 - 360 / props.length) / 2}deg);
      }
    `}
`;

interface ItemContentProps extends ItemContainerProps {
  readonly index: number;
  readonly pointingUp?: boolean | null;
}

export const PolygonalListItemContent = styled.div.attrs<ItemContentProps>(
  (props) => {
    const width = `${props.itemSize}px`;
    const top = `${(props.itemSize / 2) * -1}px`;
    const rotationDegrees =
      (360 / props.length) * (props.index + 1) + (props.pointingUp ? 45 : 225);
    return {
      style: {
        zIndex: props.zIndex,
        top,
        left: top,
        width,
        height: width,
        transform: `rotate(-${rotationDegrees}deg)`,
      },
    };
  }
)<ItemContentProps>`
  position: absolute;
`;
