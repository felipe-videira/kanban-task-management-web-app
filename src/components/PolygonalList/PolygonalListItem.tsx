import { InferProps, number } from "prop-types";
import {
  PolygonalListItemContainer as ItemContainer,
  PolygonalListItemContent as ItemContent,
} from "./styles";
import propTypes from "./propTypes";

function PolygonalListItem(
  props: InferProps<typeof PolygonalListItem.propTypes>
) {
  const zIndex = props.length - props.index;

  return (
    <ItemContainer
      zIndex={zIndex}
      length={props.length}
      itemSize={props.itemSize}
      polygonExteriorAngle={props.polygonExteriorAngle}
    >
      <ItemContent
        zIndex={zIndex}
        index={props.index}
        length={props.length}
        itemSize={props.itemSize}
        pointingUp={props.pointingUp}
        polygonExteriorAngle={props.polygonExteriorAngle}
      >
        <props.ItemComponent
          {...props.data[props.index]}
          {...props.itemProps}
        />
      </ItemContent>

      {props.index < props.length - 1 && (
        <PolygonalListItem {...props} index={props.index + 1} />
      )}
    </ItemContainer>
  );
}

PolygonalListItem.propTypes = {
  ...propTypes,
  index: number.isRequired,
  length: number.isRequired,
  polygonExteriorAngle: number.isRequired,
};

export default PolygonalListItem;
