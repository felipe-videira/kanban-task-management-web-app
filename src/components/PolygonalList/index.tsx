import { InferProps, number } from "prop-types";
import { useMemo } from "react";
import { PolygonalListContainer as Container } from "./styles";
import Item from "./PolygonalListItem";
import propTypes from "./propTypes";

function PolygonalList(props: InferProps<typeof PolygonalList.propTypes>) {
  const length = useMemo(() => props.data.length, [props.data]);
  const polygonExteriorAngle = useMemo(() => 360 / length, [length]);

  return length > 0 ? (
    <Container size={props.size} pointingUp={props.pointingUp}>
      <Item
        index={0}
        data={props.data}
        ItemComponent={props.ItemComponent}
        itemProps={props.itemProps}
        itemSize={props.itemSize}
        length={length}
        pointingUp={props.pointingUp}
        polygonExteriorAngle={polygonExteriorAngle}
      />
    </Container>
  ) : null;
}

PolygonalList.propTypes = {
  ...propTypes,
  size: number.isRequired,
};

PolygonalList.defaultProps = {
  pointingUp: true,
};

export default PolygonalList;
