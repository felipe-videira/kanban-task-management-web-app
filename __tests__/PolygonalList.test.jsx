import "jsdom-global/register";

import React from "react";
import { mount } from "enzyme";
import PolygonalList from "../src/components/PolygonalList";
import "jest-styled-components";

const mockProps = {
  data: [{ value: "value" }, { value: "value" }, { value: "value" }],
  ItemComponent: (props) => (
    // eslint-disable-next-line react/prop-types
    <button onClick={props.onClick}>{props.value}</button>
  ),
  itemProps: {
    onClick: () => {},
  },
  itemSize: 200,
  size: 400,
};

describe("PolygonalList", () => {
  it("should render correctly with all props", () => {
    const wrapper = mount(<PolygonalList {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should throw error if prop "data" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} data={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "data" is not an array of objects', () => {
    expect(() => shallow(<PolygonalList {...mockProps} data={1} />)).toThrow();
  });

  it('should throw error if prop "ItemComponent" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} ItemComponent={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "ItemComponent" is not a React element type', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} ItemComponent={() => {}} />)
    ).toThrow();
  });

  it('should throw error if prop "itemProps" is not an object', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} itemProps={"value"} />)
    ).toThrow();
  });

  it('should throw error if prop "itemSize" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} itemSize={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "itemSize" is not a number', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} itemSize={"1"} />)
    ).toThrow();
  });

  it('should throw error if prop "size" is not present', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} size={undefined} />)
    ).toThrow();
  });

  it('should throw error if prop "size" is not a number', () => {
    expect(() =>
      shallow(<PolygonalList {...mockProps} size={"1"} />)
    ).toThrow();
  });
});
