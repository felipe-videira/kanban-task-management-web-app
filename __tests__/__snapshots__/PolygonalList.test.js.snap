// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolygonalList should render correctly with all props 1`] = `
.c0 {
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 282.84542497525103px;
  height: 282.84542497525103px;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c1::after {
  content: "";
  position: absolute;
  -webkit-transform-origin: -10px -10px;
  -ms-transform-origin: -10px -10px;
  transform-origin: -10px -10px;
  height: calc(100% / (3/ 4));
  width: 20px;
  background: rgba(0,0,0,0.2);
  top: 0;
  left: 0;
  -webkit-transform: rotate(-15deg);
  -ms-transform: rotate(-15deg);
  transform: rotate(-15deg);
}

.c2 {
  position: absolute;
}

<PolygonalList
  ItemComponent={[Function]}
  data={
    Array [
      Object {
        "value": "value",
      },
      Object {
        "value": "value",
      },
      Object {
        "value": "value",
      },
    ]
  }
  itemProps={
    Object {
      "onClick": [Function],
    }
  }
  itemSize={200}
  pointingUp={true}
  size={400}
>
  <styled.div
    pointingUp={true}
    size={400}
  >
    <div
      className="c0"
      size="282.84542497525103px"
    >
      <PolygonalListItem
        ItemComponent={[Function]}
        data={
          Array [
            Object {
              "value": "value",
            },
            Object {
              "value": "value",
            },
            Object {
              "value": "value",
            },
          ]
        }
        index={0}
        itemProps={
          Object {
            "onClick": [Function],
          }
        }
        itemSize={200}
        length={3}
        pointingUp={true}
      >
        <styled.div
          itemSize={200}
          length={3}
          zIndex={3}
        >
          <div
            className="c1"
            origin="-10px"
            style={
              Object {
                "transform": "rotate(120deg)",
                "zIndex": 3,
              }
            }
          >
            <styled.div
              index={0}
              itemSize={200}
              length={3}
              pointingUp={true}
              zIndex={3}
            >
              <div
                className="c2"
                style={
                  Object {
                    "height": "200px",
                    "left": "-100px",
                    "top": "-100px",
                    "transform": "rotate(-165deg)",
                    "width": "200px",
                    "zIndex": 3,
                  }
                }
              >
                <ItemComponent
                  onClick={[Function]}
                  value="value"
                >
                  <button
                    onClick={[Function]}
                  >
                    value
                  </button>
                </ItemComponent>
              </div>
            </styled.div>
            <PolygonalListItem
              ItemComponent={[Function]}
              data={
                Array [
                  Object {
                    "value": "value",
                  },
                  Object {
                    "value": "value",
                  },
                  Object {
                    "value": "value",
                  },
                ]
              }
              index={1}
              itemProps={
                Object {
                  "onClick": [Function],
                }
              }
              itemSize={200}
              length={3}
              pointingUp={true}
            >
              <styled.div
                itemSize={200}
                length={3}
                zIndex={2}
              >
                <div
                  className="c1"
                  origin="-10px"
                  style={
                    Object {
                      "transform": "rotate(120deg)",
                      "zIndex": 2,
                    }
                  }
                >
                  <styled.div
                    index={1}
                    itemSize={200}
                    length={3}
                    pointingUp={true}
                    zIndex={2}
                  >
                    <div
                      className="c2"
                      style={
                        Object {
                          "height": "200px",
                          "left": "-100px",
                          "top": "-100px",
                          "transform": "rotate(-285deg)",
                          "width": "200px",
                          "zIndex": 2,
                        }
                      }
                    >
                      <ItemComponent
                        onClick={[Function]}
                        value="value"
                      >
                        <button
                          onClick={[Function]}
                        >
                          value
                        </button>
                      </ItemComponent>
                    </div>
                  </styled.div>
                  <PolygonalListItem
                    ItemComponent={[Function]}
                    data={
                      Array [
                        Object {
                          "value": "value",
                        },
                        Object {
                          "value": "value",
                        },
                        Object {
                          "value": "value",
                        },
                      ]
                    }
                    index={2}
                    itemProps={
                      Object {
                        "onClick": [Function],
                      }
                    }
                    itemSize={200}
                    length={3}
                    pointingUp={true}
                  >
                    <styled.div
                      itemSize={200}
                      length={3}
                      zIndex={1}
                    >
                      <div
                        className="c1"
                        origin="-10px"
                        style={
                          Object {
                            "transform": "rotate(120deg)",
                            "zIndex": 1,
                          }
                        }
                      >
                        <styled.div
                          index={2}
                          itemSize={200}
                          length={3}
                          pointingUp={true}
                          zIndex={1}
                        >
                          <div
                            className="c2"
                            style={
                              Object {
                                "height": "200px",
                                "left": "-100px",
                                "top": "-100px",
                                "transform": "rotate(-405deg)",
                                "width": "200px",
                                "zIndex": 1,
                              }
                            }
                          >
                            <ItemComponent
                              onClick={[Function]}
                              value="value"
                            >
                              <button
                                onClick={[Function]}
                              >
                                value
                              </button>
                            </ItemComponent>
                          </div>
                        </styled.div>
                      </div>
                    </styled.div>
                  </PolygonalListItem>
                </div>
              </styled.div>
            </PolygonalListItem>
          </div>
        </styled.div>
      </PolygonalListItem>
    </div>
  </styled.div>
</PolygonalList>
`;
